README.md
# Steel Crypt 3.0 (BETA)

A comprehensive library of high-level, cryptographic API's, either manually defined or pulled from PointyCastle.
This library currently supports hashing, symmetric two-way encryption, and key/IV generation. It also has 
a CLI, for conducting basic cryptography operations.

This library is a high-level wrapper over https://github.com/bcgit/pc-dart. It used to contain a fork of PointyCastle
within it, but since the package came under the ownership of the BouncyCastle organization, the additions to the fork 
have been added upstream. 

---

It takes time, effort, and mental power to keep this package updated, useful, and
improving. If you used or are using the package, I'd appreciate it if you could spare a few 
dollars to help me continue development.

[![PayPal](https://img.shields.io/static/v1?label=PayPal&message=Donate&color=blue&logo=paypal&style=for-the-badge&labelColor=black)](https://www.paypal.me/kishoredev)

---

## Null Safety

This library has been fully ported to null safety. The version 3.0.0 is identical functionally to 2.3.1+6, except for 
this porting process. There may be some slight variance in the interface, due to parameters becoming required where
they were formerly "optional" named parameters. 

---

## Note: Documentation

These docs have not been fully written yet! Steel Crypt 3.X comes with major changes, and will have a static
website on GitHub Pages for documentation. For now, you can take a look at the API reference for more
information on the beta's classes. Each class has fairly extensive documentation.

---

## Note: RSA Deprecation

As I looked at the package landscape on pub.dev, it became clear to me that packages have become oversaturated, causing
difficult decisions for independent developers. I fear contributing to this issue. 

Additionally, I haven't studied asymmetric encryption; I don't feel comfortable working with it and don't think this
package is adding much to the ecosystem as far as asymmetric encryption goes.

As of today, I will be deprecating RSA functionality in steel_crypt. It is clear to me that there are better packages
for asymmetric encryption; I highly encourage affected users to check out https://pub.dev/packages/crypton. The dev there
actively works on supporting asymmetric operations, more than I ever could, and has a deeper scope of features than this
package would have ever had.

---

[![Pub](https://img.shields.io/pub/v/steel_crypt?color=blue&label=pub&logo=Steel%20Crypt&logoColor=blue&style=for-the-badge&labelColor=black)](https://pub.dev/packages/steel_crypt)
[![License](https://img.shields.io/github/license/AKushWarrior/steel_crypt?color=blue&style=for-the-badge&labelColor=black)](https://www.mozilla.org/en-US/MPL/2.0/)
[![Commits](https://img.shields.io/github/commit-activity/m/AKushWarrior/steel_crypt?color=blue&style=for-the-badge&labelColor=black)](https://github.com/AKushWarrior/steel_crypt)

###### ©2021 Aditya Kishore
###### Licensed under the Mozilla Public License 2.0


CHANGELOG.md
# 3.0.0 - 3.0.0+1
* Port to null-safety

# 2.3.0-2.3.1+6
* Change upstream dependency to PointyCastle (instead of "pc_steelcrypt")
* Fix deprecation issues in RSACrypt
* Allow latest dependencies
* Fix CHANGELOG formatting

# 2.2.2-2.2.2+1
* Bugfix for AES-CFB and AES-OFB

# 2.2.1
* Temporarily remove tagLength while waiting for upstream fix

# 2.2.0-2.2.0+1
* Code cleanup
* Minor documentation updates
* Fix naming conflict with Stream
* Fixed a MAC bug

# 2.1.0
* Revamped AES to have new, cleaner syntax
* Cleaned up example to make sense
* Refactored code in various ways
* Fixed CMAC
* Refactored parameters to follow consistent style
* Use named parameters everywhere

# 2.0.2-2.0.3
* Syntax & formatting updates
* Documentation website nearing release
* Bugfix for MACs
* Dartdocs refined

# 2.0.1
* Fixed glitch in PBKDF2

# 2.0.0
* Breaking API changes
* Deleted unused stream ciphers
* Added Poly1305
* Added ChaCha20-Poly1305
* Added "raw" classes working with Uint8List

# 1.7.1-1.7.1+1
* Fixed critical Digest insecurity issue
* Affects anyone using a number of algorithms:
    - SHA-224, SHA-256, and SHA-1
    - RIPEMD (all versions)
    - MD4 and MD5
    - PBKDF2 with any of the above
* Affected people should update ASAP

# 1.6.1
* Fixed critical keyspace issue
* Fixed bug with RSA PEM parsing
* Updated dependencies
* Minor README updates

# 1.5.4-1.5.4+1
* Changes to random generation to adjust for base-64 bugs
* Use UTF-8 encoding more (wherever possible)
* Started work on TEA algorithm
* Updated asn1lib to placate pub

# 1.5.2-1.5.3
* RSA parsing improved
    - Added option to parse key from PEM string, not just files
    - Added option to encode key to PEM string
* Minor changes to MacCrypt
* Formatting changes for pub's new, super-strict linting
    - New casting reqs. might cause marginal slowdown
    - I'm trying to get that rating out of the dump
    - Also having to fix bugs that come with migrating a whole codebase
    - NNBD has been adopted by Pub before Dart?
    - Nothing should be broken; leave bugs in the issue tracker

# 1.5.1-1.5.1+2
* Made MacCrypt more consistent with standards
* Improved CMAC to function better
* Added GCM and GCTR to AES
* Improved README with note on prior knowledge

# 1.4.1+1
* No functional update
* Changed example to be more efficient

# 1.4.1
* Revamped ChaCha20 algorithm to:
    * Be compliant with the original definition
    * Work with the UTF-8 charset
* Bettered documentation
* Cleaned up code

# 1.3.2-1.3.2+1
* Fixed bug in ECB mode for AES
* Cleaned up code
* Updated README for AES bugfix

# 1.3.1+1
* Changed defaults for a couple classes
* Cleaned up code a little bit
* Got rid of those pesky new keywords
* Reformatted src directory
* No functional change unless you're using defaults

# 1.3.1
* Intro new algorithm: RC4
* Better documentation
* Better, more consistent README

# 1.3.0
* Introduced RSA key parsing from PEM
* ```.pubKey``` and ```.privKey``` getters are now ```.randPubKey``` and ```.randPrivKey```

# 1.2.0+2
* More major README edits
* Still no functional change

# 1.2.0+1
* Fixed major README issue
* No functional change

# 1.2.0
* Total revamp of PassCrypt()
* Scrypt password hashing now available (and default)
* PBKDF2 base algorithm now changeable
* Lowered PBKDF2 rounds to 10000 to increase speed

# 1.1.2
* New padding option: ISO10126-2
* Security fix to X9.23 padding

# 1.1.1-1.1.1+2
* Various bug fixes regarding static variables
* README edits for clarity

# 1.1.0
* Added Grain-128 algorithm
* Added ISAAC algorithm
* README updates


# 1.0.1-1.0.1+3
* Fixed breaking bug in CMAC
* Added HC-256 algorithm
* README updates

# 1.0.0
* Release version! This package is now stable and tested.
* Old .hashHMAC class now obsolete
* Added new MacCrypt() class with CMAC and HMAC
* Added two new padding options for AES

# 0.8.2+1
* Minor updates to pubspec
* No functional change

# 0.8.2
* New reduced-round variants of Salsa20
* Better documentation
* Minor bug fixes to hashing classes

# 0.8.1-0.8.1+5
* Revamped AES with new padding and better explanation in README
* Made HashCrypt and PassCrypt less buggy to use
* Changed pointycastle dependency to github in order to get new padding
* Phased out encrypt package
* Added two new variants of ChaCha20
* Bettered pub code health

# 0.7.1+2
* Minor Readme update
* No functional change

# 0.7.1+1
* Minor Readme update
* No functional change

# 0.7.1
* Revamped KeyCrypt() to have two methods
* One generating with Fortuna
* Other generating with Random.secure()
* Old key generation method no longer works

# 0.6.2
* Made it so HashCrypt() and PassCrypt() work with all strings
* Updated AES documentation in README to reflect different modes

# 0.6.1+1
* Hotfix for breaking bug regarding meta package
* Fixes ChaCha20

# 0.6.1
* Reorganization of files
* Implemented ChaCha20
* Changed SymCrypt() to AesCrypt()
* Moved Salsa20 and ChaCha20 to new LightCrypt() class
* Ditched Threefish due to lack of demand

# 0.5.5+1
* Enforce IV uniqueness in SymCrypt class
* Minor updates to README
* Work begins in threefish.dart

# 0.5.5
* Changed IV + Key generation to Fortuna for increased speed + entropy
* Increases security, but no front-end changes
* Improved runtime efficiency by declaring top-end variables and helper functions static
* Decreased PassCrypt rounds from 20000 to 15000 to decrease runtime 

# 0.5.4+1
* Added Mozilla License notice + copyright to every code file
* Included copyright at bottom of README
* Updated README Notes + TODO's
* No functional improvement

# 0.5.4
* Fixing hash funct. of CLI

# 0.5.3+4
* Disregard, README isn't processing properly
* No functional improvement

# 0.5.3+3
* More README fixes
* Still no functional improvement

# 0.5.3+2
* README looks terrible on pub, patched
* No functional improvement

# 0.5.3+1
* Edited README to illustrate best usage of CLI
* No functional improvement

# 0.5.3
* Added new CLI to increase ease of use

# 0.5.2+1

lib/src/light.dart
//This Source Code Form is subject to the terms of the Mozilla Public
//License, v. 2.0. If a copy of the MPL was not distributed with this
//file, You can obtain one at https://mozilla.org/MPL/2.0/.

// © 2019 Aditya Kishore

part of 'steel_crypt_base.dart';

///Class containing stream ciphers of every kind.
class LightCrypt {
  core.String _type;
  core.String _key;
  StreamCipher _encrypter;

  ///Get name of this LightCrypt's algorithm.
  String get algorithm {
    return _type;
  }

  ///Get this LightCrypt's key.
  String get key {
    return _key;
  }

  ///Construct encryption machine using key and algorithm.
  LightCrypt(core.String key, [core.String algorithm = 'ChaCha20']) {
    _type = algorithm;
    _key = key;
    if (_type == 'Salsa20' ||
        _type == 'Salsa20/8' ||
        _type == 'Salsa20/12' ||
        _type == 'ChaCha20' ||
        _type == 'ChaCha20/12' ||
        _type == 'ChaCha20/8') {
      _encrypter = StreamCipher(_type);
    } else if (_type == 'Salsa20/20') {
      _type = 'Salsa20';
      _encrypter = StreamCipher(_type);
    } else if (_type == 'ChaCha20/20') {
      _type = 'ChaCha20';
      _encrypter = StreamCipher(_type);
    } else if (_type == 'HC-256' ||
        _type == 'Grain-128' ||
        _type == 'ISAAC' ||
        _type == 'RC4') {
    } else {
      throw ArgumentError(
          "This algorithm isn't supported. Check for typos, or file a feature request.");
    }
  }

  ///Encrypt (with iv) and return in base 64
  core.String encrypt(core.String input, core.String iv) {
    if (_type == 'Salsa20' ||
        _type == 'Salsa20/8' ||
        _type == 'Salsa20/12' ||
        _type == 'ChaCha20' ||
        _type == 'ChaCha20/12' ||
        _type == 'ChaCha20/8') {
      var localKey = Uint8List.fromList(key.codeUnits);
      var localIV = Uint8List.fromList(iv.codeUnits);
      var localInput = utf8.encode(input);
      var params = ParametersWithIV<KeyParameter>(
          KeyParameter(Uint8List.fromList(localKey.sublist(0, 32))),
          Uint8List.fromList(localIV.sublist(0, 8)));
      _encrypter..init(true, params);
      var inter = _encrypter.process(Uint8List.fromList(localInput));
      return base64.encode(inter);
    } else if (_type == 'Grain-128') {
      var machine = StreamCipher(_type);
      var localKey = Uint8List.fromList(key.codeUnits);
      var localIV = Uint8List.fromList(iv.codeUnits);
      var localInput = utf8.encode(input);
      var params = ParametersWithIV<KeyParameter>(
          KeyParameter(Uint8List.fromList(localKey.sublist(0, 32))),
          Uint8List.fromList(localIV.sublist(0, 12)));
      machine..init(false, params);
      var inter = machine.process(Uint8List.fromList(localInput));
      return base64.encode(inter);
    } else if (_type == 'ISAAC' || _type == 'RC4') {
      var machine = StreamCipher(_type);
      var localKey = Uint8List.fromList(key.codeUnits);
      var localInput = utf8.encode(input);
      var params = KeyParameter(Uint8List.fromList(localKey.sublist(0, 32)));
      machine..init(false, params);
      var inter = machine.process(Uint8List.fromList(localInput));
      return base64.encode(inter);
    } else if (_type == 'HC-256') {
      var machine = StreamCipher(_type);
      var localKey = Uint8List.fromList(key.codeUnits);
      var localIV = Uint8List.fromList(iv.codeUnits);
      var localInput = utf8.encode(input);
      var params = ParametersWithIV<KeyParameter>(
          KeyParameter(Uint8List.fromList(localKey.sublist(0, 32))),
          Uint8List.fromList(localIV.sublist(0, 16)));
      machine..init(false, params);
      var inter = machine.process(Uint8List.fromList(localInput));
      return base64.encode(inter);
    }
    return '';
  }

  ///Decrypt base 64 (with iv) and return original
  core.String decrypt(core.String encrypted, core.String iv) {
    if (_type == 'Salsa20' ||
        _type == 'Salsa20/8' ||
        _type == 'Salsa20/12' ||
        _type == 'ChaCha20' ||
        _type == 'ChaCha20/12' ||
        _type == 'ChaCha20/8') {
      var localKey = Uint8List.fromList(key.codeUnits);
      var localIV = Uint8List.fromList(iv.codeUnits);
      var localInput = base64.decode(encrypted);
      var params = ParametersWithIV<KeyParameter>(
          KeyParameter(Uint8List.fromList(localKey.sublist(0, 32))),
          Uint8List.fromList(localIV.sublist(0, 8)));
      _encrypter..init(false, params);
      var inter = _encrypter.process(localInput);
      return utf8.decode(inter);
    } else if (_type == 'Grain-128') {
      var machine = StreamCipher(_type);
      var localKey = Uint8List.fromList(key.codeUnits);
      var localIV = Uint8List.fromList(iv.codeUnits);
      var localInput = base64.decode(encrypted);
      var params = ParametersWithIV<KeyParameter>(
          KeyParameter(Uint8List.fromList(localKey.sublist(0, 32))),
          Uint8List.fromList(localIV.sublist(0, 12)));
      machine..init(false, params);
      var inter = machine.process(localInput);
      return utf8.decode(inter);
    } else if (_type == 'HC-256') {
      var machine = StreamCipher(_type);
      var localKey = Uint8List.fromList(key.codeUnits);
      var localIV = Uint8List.fromList(iv.codeUnits);
      var localInput = base64.decode(encrypted);
      var params = ParametersWithIV<KeyParameter>(
          KeyParameter(Uint8List.fromList(localKey.sublist(0, 32))),
          Uint8List.fromList(localIV.sublist(0, 16)));
      machine..init(false, params);
      var inter = machine.process(localInput);
      return utf8.decode(inter);
    } else if (_type == 'ISAAC' || _type == 'RC4') {
      var machine = StreamCipher(_type);
      var localKey = Uint8List.fromList(key.codeUnits);
      var localInput = base64.decode(encrypted);
      var params = KeyParameter(Uint8List.fromList(localKey.sublist(0, 32)));
      machine..init(false, params);
      var inter = machine.process(localInput);
      return utf8.decode(inter);
    }
    return '';
  }
}
